pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,345 @@
+
+    .. note::
+        This transform will raise an error if it encounters non-commuting terms. To avoid non-commuting terms in
+        circuit measurements, the :func:`split_non_commuting <pennylane.transforms.split_non_commuting>` transform
+        can be applied.
+
+
+    **Examples:**
+
+    This transform allows us to transform QNode measurements into the measurement basis by adding
+    the relevant diagonalizing gates to the end of the tape operations.
+
+    .. code-block:: python3
+
+        from pennylane.transforms import diagonalize_measurements
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @diagonalize_measurements
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.RY(x[0], wires=0)
+            qml.RX(x[1], wires=1)
+            return qml.expval(qml.X(0) @ qml.Z(1)), qml.var(0.5 * qml.Y(2) + qml.X(0))
+
+    Instead of decorating the QNode, we can also create a new function that yields the same
+    result in the following way:
+
+    .. code-block:: python3
+
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.RY(x[0], wires=0)
+            qml.RX(x[1], wires=1)
+            return qml.expval(qml.X(0) @ qml.Z(1)), qml.var(0.5 * qml.Y(2) + qml.X(0))
+
+        diagonalized_circuit = diagonalize_measurements(circuit)
+
+    Applying the transform appends the relevant gates to the end of the cirucit to allow
+    measurements to be in the Z basis, so the original circuit
+
+    >>> print(qml.draw(circuit)([np.pi/4, np.pi/4]))
+    0: ──RY(0.79)─┤ ╭<X@Z> ╭Var[(0.50*Y)+X]
+    1: ──RX(0.79)─┤ ╰<X@Z> │
+    2: ───────────┤        ╰Var[(0.50*Y)+X]
+
+    becomes
+
+    >>> print(qml.draw(diagonalized_circuit)([np.pi/4, np.pi/4]))
+    0: ──RY(0.79)──H────┤ ╭<Z@Z> ╭Var[(0.50*Z)+Z]
+    1: ──RX(0.79)───────┤ ╰<Z@Z> │
+    2: ──Z─────────S──H─┤        ╰Var[(0.50*Z)+Z]
+
+    >>> diagonalized_circuit([np.pi/4, np.pi/4])
+    (tensor(0.5, requires_grad=True), tensor(0.75, requires_grad=True))
+
+    .. details::
+        :title: Usage Details
+
+        The transform can also diagonalize only a subset of the operators. Be default, the only
+        supported base observable is PauliZ. What if a backend device can handle
+        X, Y and Z, but doesn't provide support for Hadamard? We can set this by passing
+        ``supported_base_obs`` to the transform. Let's create a tape with some measurements:
+
+        .. code-block:: python3
+
+            measurements = [
+                qml.expval(qml.X(0) + qml.Hadamard(1)),
+                qml.expval(qml.X(0) + 0.2 * qml.Hadamard(1)),
+                qml.var(qml.Y(2) + qml.X(0)),
+            ]
+            tape = qml.tape.QuantumScript(measurements=measurements)
+            tapes, processing_fn = diagonalize_measurements(tape,
+                                                                 supported_base_obs=['PauliX', 'PauliY', 'PauliZ'])
```

Comment:
```suggestion
            tapes, processing_fn = diagonalize_measurements(
                tape,
                supported_base_obs=['PauliX', 'PauliY', 'PauliZ']
            )
```
Making it fully multi-line should help with readability.